sch_cake.c

```diff
@@ -1637,28 +1665,50 @@ static u8 cake_handle_diffserv(struct sk_buff *skb, u16 wash)
 {
 	int wlen = skb_network_offset(skb);
 	u8 dscp;
+	unsigned char buf[2];
+	struct ipv6hdr *ip6h;
+	struct iphdr *iph;
 
 	switch (cake_skb_proto(skb)) {
 	case htons(ETH_P_IP):
-		wlen += sizeof(struct iphdr);
-		if (!pskb_may_pull(skb, wlen) ||
-		    skb_try_make_writable(skb, wlen))
+		/* ToS is in the second byte of iphdr */
+		iph = skb_header_pointer(skb, wlen, sizeof(buf), &buf);
+
+		if (unlikely(iph == NULL))
 			return 0;
 
-		dscp = ipv4_get_dsfield(ip_hdr(skb)) >> 2;
-		if (wash && dscp)
+		dscp = ipv4_get_dsfield(iph) >> 2;
+
+		if (wash && dscp) {
+			wlen += sizeof(struct iphdr);
```

Comment:
...and then (with the rename above) redefine this as a scope-local variable computed from offset+sizeof()